File: OpenBurnBarMobile/Services/ChartStudio/TrendDataDigest.swift
import Foundation
import OpenBurnBarCore

// MARK: - Trend Data Digest
//
// Compact, JSON-serializable snapshot of everything Hermes needs to draw
// charts about the user's burn. Capped at ~6KB so it always fits in a
// single Hermes turn even on small open-source models.
//
// This type is `Sendable` and pure-value so it can be built off the main
// thread and unit-tested without view dependencies.

public struct TrendDataDigest: Codable, Hashable, Sendable {

    // MARK: - Top-level totals

    public struct WindowTotals: Codable, Hashable, Sendable {
        public let window: String          // "today" | "7d" | "30d"
        public let costUsd: Double
        public let tokens: Int
        public let requests: Int
    }

    // MARK: - Per-provider summary

    public struct ProviderSlice: Codable, Hashable, Sendable {
        public let provider: String        // display name e.g. "Claude Code"
        public let providerKey: String     // persistedToken
        public let costUsd: Double
        public let tokens: Int
        public let requests: Int
        public let sharePct: Double        // 0...100
    }

    // MARK: - Per-model summary

    public struct ModelSlice: Codable, Hashable, Sendable {
        public let model: String
        public let provider: String
        public let costUsd: Double
        public let tokens: Int
        public let requests: Int
        public let sharePct: Double
    }

    // MARK: - Per-project summary

    public struct ProjectSlice: Codable, Hashable, Sendable {
        public let project: String
        public let costUsd: Double
        public let tokens: Int
        public let sessions: Int
    }

    // MARK: - Per-device summary

    public struct DeviceSlice: Codable, Hashable, Sendable {
        public let device: String
        public let tokens: Int
        public let requests: Int
    }

    // MARK: - Daily series

    public struct DailySeries: Codable, Hashable, Sendable {
        public let date: String            // ISO `yyyy-MM-dd`
        public let total: Double
        public let perProvider: [String: Double]   // providerKey → value
    }

    // MARK: - Hour-of-day histogram

    public struct HourBucket: Codable, Hashable, Sendable {
        public let hour: Int               // 0...23
        public let costUsd: Double
        public let tokens: Int
    }

    // MARK: - Recent session

    public struct SessionSlice: Codable, Hashable, Sendable {
        public let id: String
        public let startedAt: String       // ISO 8601
        public let durationSec: Int
        public let provider: String
        public let providerKey: String
        public let model: String
        public let project: String
        public let inputTokens: Int
        public let outputTokens: Int
        public let cacheReadTokens: Int
        public let cacheCreationTokens: Int
        public let reasoningTokens: Int
        public let costUsd: Double
        public let cacheHitRate: Double    // 0...1
        public let outputTokensPerSecond: Double
    }

    // MARK: - Cache aggregates

    public struct CacheAggregate: Codable, Hashable, Sendable {
        public let totalCacheReadTokens: Int
        public let totalCacheCreationTokens: Int
        public let totalInputTokens: Int
        public let cacheHitRate: Double                // 0...1 across window
        public let estSavingsUsd: Double               // rough: cacheRead × (input price - cache read price), see prompt
    }

    // MARK: - Display mode hint

    public let displayMode: String                     // "currency" | "tokens"
    public let generatedAt: String                     // ISO 8601
    public let windowDescription: String               // human label e.g. "last 30 days"

    public let totals: [WindowTotals]
    public let providers: [ProviderSlice]
    public let models: [ModelSlice]
    public let projects: [ProjectSlice]
    public let devices: [DeviceSlice]
    public let daily: [DailySeries]
    public let hourly: [HourBucket]
    public let recentSessions: [SessionSlice]
    public let cache: CacheAggregate

    // MARK: - Builder

    /// Build a digest from `DashboardStore` + `ActivityStore` data. Caps each
    /// list to a sensible size so the JSON stays under ~6KB.
    public static func build(
        windowTotals: [RollupWindowKey: RollupTotals],
        providerSummaries: [RollupProviderSummary],
        modelSummaries: [RollupModelSummary],
        deviceSummaries: [RollupDeviceSummary],
        dailyPoints: [RollupDailyPoint],
        recentUsages: [TokenUsage],
        displayMode: UsageDisplayMode,
        windowDescription: String = "last 30 days",
        now: Date = Date()
    ) -> TrendDataDigest {

        // Totals (today / 7d / 30d only — keep it tight)
        let totals: [WindowTotals] = [
            (RollupWindowKey.today, "today"),
            (RollupWindowKey.sevenDays, "7d"),
            (RollupWindowKey.thirtyDays, "30d")
        ].compactMap { key, label in
            guard let t = windowTotals[key] else { return nil }
            return WindowTotals(
                window: label,
                costUsd: t.costUsd,
                tokens: t.tokens,
                requests: t.requests
            )
        }

        // Providers (top 6, with share %)
        let providerTotal = providerSummaries.reduce(0.0) { $0 + Double($1.totalTokens) }
        let providerSlices = providerSummaries.prefix(6).map { p -> ProviderSlice in
            let share = providerTotal > 0
                ? (Double(p.totalTokens) / providerTotal) * 100
                : 0
            let displayProvider = AgentProvider.fromPersistedToken(p.provider)
                ?? AgentProvider.fromProviderID(p.providerID)
            return ProviderSlice(
                provider: displayProvider?.displayName ?? p.provider,
                providerKey: p.provider,
                costUsd: p.totalCost ?? 0,
                tokens: p.totalTokens,
                requests: p.totalRequests,
                sharePct: share
            )
        }

        // Models (top 8, with share %)
        let modelTotal = modelSummaries.reduce(0) { $0 + $1.tokens }
        let modelSlices = modelSummaries.prefix(8).map { m -> ModelSlice in
            let share = modelTotal > 0
                ? (Double(m.tokens) / Double(modelTotal)) * 100
                : 0
            return ModelSlice(
                model: m.model,
                provider: m.provider,
                costUsd: m.cost ?? 0,
                tokens: m.tokens,
                requests: m.requests,
                sharePct: share
            )
        }

        // Devices (top 4)
        let deviceSlices = deviceSummaries.prefix(4).map {
            DeviceSlice(
                device: $0.deviceId,
                tokens: $0.tokens,
                requests: $0.requests
            )
        }

        // Daily series — bucket recent usage by day to derive perProvider stack
        let calendar = Calendar(identifier: .gregorian)
        let daily = buildDailySeries(
            dailyPoints: dailyPoints,
            recentUsages: recentUsages,
            calendar: calendar
        )

        // Hour-of-day histogram from recentUsages (last 14 days)
        let hourly = buildHourly(
            recentUsages: recentUsages,
            calendar: calendar,
            now: now
        )

        // Project slices — group recentUsages by projectName, top 6
        let projectGroups = Dictionary(grouping: recentUsages) { $0.projectName }
        let projectSlices = projectGroups
            .map { project, items -> ProjectSlice in
                let cost = items.reduce(0.0) { $0 + $1.cost }
                let tokens = items.reduce(0) { $0 + $1.totalTokens }
                let sessions = Set(items.map(\.sessionId)).count
                return ProjectSlice(
                    project: project.isEmpty ? "Unknown" : project,
                    costUsd: cost,
                    tokens: tokens,
                    sessions: sessions
                )
            }
            .sorted { $0.costUsd > $1.costUsd }
            .prefix(6)
            .map { $0 }

        // Recent sessions (top 25 by start time)
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime]
        let recentSessions = recentUsages.prefix(25).map { u -> SessionSlice in
            let duration = Int(max(0, u.endTime.timeIntervalSince(u.startTime)))
            let promptBasis = max(0, u.inputTokens) + max(0, u.cacheCreationTokens) + max(0, u.cacheReadTokens)
            let cacheHit: Double = promptBasis > 0
                ? Double(max(0, u.cacheReadTokens)) / Double(promptBasis)
                : 0
            let velocity: Double = duration > 0
                ? Double(u.outputTokens) / Double(duration)
                : 0
            return SessionSlice(
                id: u.sessionId,
                startedAt: iso.string(from: u.startTime),
                durationSec: duration,
                provider: u.provider.displayName,
                providerKey: u.provider.persistedToken,
                model: u.model,
                project: u.projectName.isEmpty ? "Unknown" : u.projectName,
                inputTokens: u.inputTokens,
                outputTokens: u.outputTokens,
                cacheReadTokens: u.cacheReadTokens,
                cacheCreationTokens: u.cacheCreationTokens,
                reasoningTokens: u.reasoningTokens,
                costUsd: u.cost,
                cacheHitRate: cacheHit,
                outputTokensPerSecond: velocity
            )
        }

        // Cache aggregate
        let totalInput = recentUsages.reduce(0) { $0 + max(0, $1.inputTokens) }
        let totalCacheRead = recentUsages.reduce(0) { $0 + max(0, $1.cacheReadTokens) }
        let totalCacheCreation = recentUsages.reduce(0) { $0 + max(0, $1.cacheCreationTokens) }
        let basis = totalInput + totalCacheRead + totalCacheCreation
        let cacheHitRate: Double = basis > 0 ? Double(totalCacheRead) / Double(basis) : 0
        // Rough savings estimate: assume non-cached input is 4× the cost of cache read
        // (Anthropic's published 1× vs 0.25× → 75% cheaper). This is a heuristic
        // shown to Hermes; it is *not* user-facing currency without disclaimer.
        let estSavings = Double(totalCacheRead) * 0.000003
        let cache = CacheAggregate(
            totalCacheReadTokens: totalCacheRead,
            totalCacheCreationTokens: totalCacheCreation,
            totalInputTokens: totalInput,
            cacheHitRate: cacheHitRate,
            estSavingsUsd: estSavings
        )

        return TrendDataDigest(
            displayMode: displayMode.rawValue,
            generatedAt: iso.string(from: now),
            windowDescription: windowDescription,
            totals: totals,
            providers: Array(providerSlices),
            models: Array(modelSlices),
            projects: projectSlices,
            devices: Array(deviceSlices),
            daily: daily,
            hourly: hourly,
            recentSessions: Array(recentSessions),
            cache: cache
        )
    }

    /// Encodes the digest as compact JSON. Used both by the prompt engine
    /// and by debug overlays. Returns empty string on encoder failure.
    public func compactJSON() -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        guard let data = try? encoder.encode(self) else { return "{}" }
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    /// Approximate byte size of the JSON payload — used by tests to keep us
    /// under the 6KB ceiling.
    public var approximateByteSize: Int {
        compactJSON().utf8.count
    }

    // MARK: - Daily series helpers

    private static func buildDailySeries(
        dailyPoints: [RollupDailyPoint],
        recentUsages: [TokenUsage],
        calendar: Calendar
    ) -> [DailySeries] {
        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "UTC")
        formatter.dateFormat = "yyyy-MM-dd"

        // Bucket recent usage by day → providerKey → value (cost)
        var perDay: [String: [String: Double]] = [:]
        for usage in recentUsages {
            let day = formatter.string(from: usage.startTime)
            perDay[day, default: [:]][usage.provider.persistedToken, default: 0] += usage.cost
        }

        // Cap to last 30 daily points to control payload size
        return dailyPoints.suffix(30).map { point in
            let day = formatter.string(from: point.date)
            return DailySeries(
                date: day,
                total: point.value,
                perProvider: perDay[day] ?? [:]
            )
        }
    }

    private static func buildHourly(
        recentUsages: [TokenUsage],
        calendar: Calendar,
        now: Date
    ) -> [HourBucket] {
        let cutoff = calendar.date(byAdding: .day, value: -14, to: now) ?? now
        var buckets: [Int: (Double, Int)] = [:]
        for usage in recentUsages where usage.startTime >= cutoff {
            let hour = calendar.component(.hour, from: usage.startTime)
            let existing = buckets[hour] ?? (0, 0)
            buckets[hour] = (existing.0 + usage.cost, existing.1 + usage.totalTokens)
        }
        return (0..<24).map { h in
            let (cost, tokens) = buckets[h] ?? (0, 0)
            return HourBucket(hour: h, costUsd: cost, tokens: tokens)
        }
    }
}

File: android/app/src/main/java/com/openburnbar/data/derived/TrendDataDigest.kt
package com.openburnbar.data.derived

import com.openburnbar.data.models.AgentProvider
import com.openburnbar.data.models.LLMModelBrand
import com.openburnbar.data.models.RollupSummary
import com.openburnbar.data.models.TokenUsage
import com.openburnbar.data.models.UsageDisplayMode
import com.openburnbar.data.models.UsageRollups
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlin.math.max

internal object TrendDigestConstants {
    const val DEFAULT_SEMANTIC_HASH_LIMIT = 24
    const val PERCENT_SCALE = 1_000_000_000
    const val PERCENT_SCALE_DOUBLE = 1_000_000_000.0
    const val PERCENT_SCALE_MILLION = 1_000_000
    const val MIN_COST_DELTA_USD = 0.01
    const val CACHE_RATE_HIGH_THRESHOLD = 0.5
    const val CACHE_RATE_LOW_THRESHOLD = 0.25
    const val SAVINGS_COST_FACTOR = 0.9
    const val AVG_INPUT_COST_USD_PER_MILLION_TOKENS = 3.0
    const val PROVIDER_SHARE_DOMINANT_THRESHOLD = 50.0
    const val WEEK_OVER_WEEK_ALERT_THRESHOLD_PERCENT = 5.0
    const val HIGH_TOKEN_VOLUME_THRESHOLD = 100_000
    const val MIN_SESSION_DURATION_SECONDS = 5.0
    const val HOURS_PER_DAY = 24
    const val SECONDS_PER_MINUTE = 60
    const val MILLIS_PER_SECOND = 1000L
    const val MAX_HOUR_OF_DAY = 23
    const val ROLLING_WINDOW_DAYS = 7.0
}

/**
 * Compact, immutable analytics snapshot consumed by the Trend Atlas card and
 * insight rotator. Pure-Kotlin port of the iOS `TrendDataDigest` — same shape,
 * same caps, computed locally from the data already on the device so no
 * Firestore schema work is needed.
 */
data class TrendDataDigest(
    val displayMode: UsageDisplayMode,
    val generatedAtMs: Long,
    val windowDescription: String,
    val totals: List<WindowTotals>,
    val providers: List<ProviderSlice>,
    val models: List<ModelSlice>,
    val projects: List<ProjectSlice>,
    val devices: List<DeviceSlice>,
    val daily: List<DailySeries>,
    val hourly: List<HourBucket>,
    val recentSessions: List<SessionSlice>,
    val cache: CacheAggregate,
) {
    data class WindowTotals(
        // "today" | "7d" | "30d"
        val window: String,
        val costUsd: Double,
        val tokens: Long,
        val requests: Int,
    )

    data class ProviderSlice(
        // display name
        val provider: String,
        // raw persisted token
        val providerKey: String,
        val costUsd: Double,
        val tokens: Long,
        val requests: Int,
        // 0..100
        val sharePct: Double,
    )

    data class ModelSlice(
        val model: String,
        val provider: String,
        val providerKey: String,
        val brand: LLMModelBrand,
        val costUsd: Double,
        val tokens: Long,
        val requests: Int,
        val sharePct: Double,
    )

    data class ProjectSlice(
        val project: String,
        val costUsd: Double,
        val tokens: Long,
        val sessions: Int,
    )

    data class DeviceSlice(
        val device: String,
        val tokens: Long,
        val requests: Int,
    )

    data class DailySeries(
        // ISO yyyy-MM-dd
        val date: String,
        val total: Double,
        // providerKey → cost
        val perProvider: Map<String, Double>,
    )

    data class HourBucket(
        // 0..23
        val hour: Int,
        val costUsd: Double,
        val tokens: Long,
    )

    data class SessionSlice(
        val id: String,
        val startedAtMs: Long,
        val durationSec: Int,
        val provider: String,
        val providerKey: String,
        val model: String,
        val project: String,
        val inputTokens: Int,
        val outputTokens: Int,
        val cacheReadTokens: Int,
        val cacheCreationTokens: Int,
        val reasoningTokens: Int,
        val costUsd: Double,
        // 0..1
        val cacheHitRate: Double,
        val outputTokensPerSecond: Double,
    )

    data class CacheAggregate(
        val totalCacheReadTokens: Long,
        val totalCacheCreationTokens: Long,
        val totalInputTokens: Long,
        // 0..1 across window
        val cacheHitRate: Double,
        val estSavingsUsd: Double,
    )

    companion object {
        private const val MAX_DAILY_DAYS = 21
        private const val MAX_HOURLY_LOOKBACK_DAYS = 14
        private const val MAX_PROVIDERS = 6
        private const val MAX_MODELS = 8
        private const val MAX_PROJECTS = 5
        private const val MAX_DEVICES = 4
        private const val MAX_SESSIONS = 15
        private const val MAX_PROVIDERS_PER_DAY = 4

        /**
         * Build the digest from the raw streams already flowing into Pulse.
         * `recentUsages` should be sorted desc by timestamp — we never order
         * them ourselves so the caller's existing pagination semantics win.
         */
        fun build(
            rollups: UsageRollups,
            recentUsages: List<TokenUsage>,
            displayMode: UsageDisplayMode = UsageDisplayMode.CURRENCY,
            windowDescription: String = "last 30 days",
            now: Date = Date(),
        ): TrendDataDigest {
            val totals =
                listOf(
                    WindowTotals(
                        window = "today",
                        costUsd = rollups.today,
                        tokens = rollups.todayTokens,
                        requests = countRequests(recentUsages, withinDays = 1, now = now),
                    ),
                    WindowTotals(
                        window = "7d",
                        costUsd = rollups.sevenDays,
                        tokens = rollups.sevenDayTokens,
                        requests = countRequests(recentUsages, withinDays = 7, now = now),
                    ),
                    WindowTotals(
                        window = "30d",
                        costUsd = rollups.thirtyDays,
                        tokens = rollups.thirtyDayTokens,
                        requests = countRequests(recentUsages, withinDays = 30, now = now),
                    ),
                )

            val providers = buildProviders(rollups.providerSummaries)
            val models = buildModels(rollups.modelSummaries)
            val projects = buildProjects(recentUsages)
            val devices = buildDevices(recentUsages)
            val daily = buildDailySeries(rollups.dailyPoints, recentUsages)
            val hourly = buildHourly(recentUsages, now)
            val sessions = buildSessions(recentUsages)
            val cache = trendDigestBuildCache(recentUsages)

            return TrendDataDigest(
                displayMode = displayMode,
                generatedAtMs = now.time,
                windowDescription = windowDescription,
                totals = totals,
                providers = providers,
                models = models,
                projects = projects,
                devices = devices,
                daily = daily,
                hourly = hourly,
                recentSessions = sessions,
                cache = cache,
            )
        }

        // ── Builders ──

        private fun countRequests(usages: List<TokenUsage>, withinDays: Int, now: Date): Int {
            val cutoff =
                now.time -
                    withinDays.toLong() *
                    TrendDigestConstants.HOURS_PER_DAY *
                    TrendDigestConstants.SECONDS_PER_MINUTE *
                    TrendDigestConstants.SECONDS_PER_MINUTE *
                    TrendDigestConstants.MILLIS_PER_SECOND
            return usages.count { it.timestamp >= cutoff }
        }

        private fun buildProviders(summaries: List<RollupSummary>): List<ProviderSlice> {
            // Collapse summaries by provider so duplicate account rows fold
            // together — the per-account view lives elsewhere.
            val byProvider =
                summaries.groupBy { it.provider }
                    .map { (key, group) ->
                        RollupSummary(
                            provider = key,
                            providerId = group.first().providerId,
                            accountLabel = "",
                            totalRequests = group.sumOf { it.totalRequests },
                            totalTokens = group.sumOf { it.totalTokens },
                            totalCost = group.sumOf { it.totalCost },
                        )
                    }
                    .sortedByDescending { it.totalCost }

            val total = byProvider.sumOf { it.totalTokens }.toDouble().coerceAtLeast(1.0)
            return byProvider.take(MAX_PROVIDERS).map { p ->
                val agent = AgentProvider.fromKey(p.provider)
                ProviderSlice(
                    provider = agent?.displayName ?: p.provider,
                    providerKey = agent?.key ?: p.provider,
                    costUsd = p.totalCost,
                    tokens = p.totalTokens,
                    requests = p.totalRequests,
                    sharePct = p.totalTokens.toDouble() / total * 100.0,
                )
            }
        }

        private fun buildModels(summaries: List<RollupSummary>): List<ModelSlice> {
            val total = summaries.sumOf { it.totalTokens }.toDouble().coerceAtLeast(1.0)
            // RollupSummary doesn't carry the model name as its own field — it
            // lives in `accountLabel` for model summaries (mirrors iOS where
            // ModelSlice flattens through the same RollupSummary struct).
            return summaries
                .sortedByDescending { it.totalCost }
                .take(MAX_MODELS)
                .map { m ->
                    val modelName = m.accountLabel.ifBlank { m.provider }
                    ModelSlice(
                        model = modelName,
                        provider = AgentProvider.fromKey(m.provider)?.displayName ?: m.provider,
                        providerKey = m.provider,
                        brand = LLMModelBrand.infer(modelName),
                        costUsd = m.totalCost,
                        tokens = m.totalTokens,
                        requests = m.totalRequests,
                        sharePct = m.totalTokens.toDouble() / total * 100.0,
                    )
                }
        }

        private fun buildProjects(usages: List<TokenUsage>): List<ProjectSlice> {
            return usages
                .mapNotNull { usage ->
                    usage.projectName?.takeIf { it.isNotBlank() }?.let { projectName -> projectName to usage }
                }
                .groupBy(keySelector = { it.first }, valueTransform = { it.second })
                .map { (project, list) ->
                    ProjectSlice(
                        project = project,
                        costUsd = list.sumOf { it.effectiveCost },
                        tokens = list.sumOf { it.totalTokens.toLong() },
                        sessions = list.distinctBy { it.sessionId ?: it.id }.size,
                    )
                }
                .sortedByDescending { it.costUsd }
                .take(MAX_PROJECTS)
        }

        private fun buildDevices(usages: List<TokenUsage>): List<DeviceSlice> {
            return usages
                .mapNotNull { usage ->
                    val deviceId =
                        usage.deviceId?.takeIf { it.isNotBlank() }
                            ?: usage.sourceDeviceId?.takeIf { it.isNotBlank() }
                    deviceId?.let { it to usage }
                }
                .groupBy(keySelector = { it.first }, valueTransform = { it.second })
                .map { (device, list) ->
                    DeviceSlice(
                        device = device,
                        tokens = list.sumOf { it.totalTokens.toLong() },
                        requests = list.size,
                    )
                }
                .sortedByDescending { it.tokens }
                .take(MAX_DEVICES)
        }

        private fun buildDailySeries(dailyPoints: Map<String, Double>, usages: List<TokenUsage>): List<DailySeries> {
            // Group usages by day-of-timestamp in user's local TZ so the
            // x-axis lines up with what the user sees on the calendar.
            val cal = Calendar.getInstance()
            val dayKey: (Long) -> String = { ms ->
                cal.time = Date(ms)
                cal.set(Calendar.HOUR_OF_DAY, 0)
                cal.set(Calendar.MINUTE, 0)
                cal.set(Calendar.SECOND, 0)
                cal.set(Calendar.MILLISECOND, 0)
                ISO_DAY.format(cal.time)
            }

            val perDayProviderTotals = mutableMapOf<String, MutableMap<String, Double>>()
            for (u in usages) {
                val date = dayKey(u.timestamp)
                val providerKey = AgentProvider.fromKey(u.provider)?.key ?: u.provider
                val bucket = perDayProviderTotals.getOrPut(date) { mutableMapOf() }
                bucket.merge(providerKey, u.effectiveCost) { a, b -> a + b }
            }

            // Union all day keys from dailyPoints (server-side totals) AND
            // any days we found locally. dailyPoints supplies fallback totals
            // for days that aren't in the recent-usages window.
            val allDays =
                (dailyPoints.keys + perDayProviderTotals.keys)
                    .toSortedSet()
                    .toList()
                    .takeLast(MAX_DAILY_DAYS)

            return allDays.map { date ->
                val perProvider = perDayProviderTotals[date] ?: emptyMap()
                // Keep only the top providers per day so the JSON / stacks
                // don't explode when a user has 10+ active providers.
                val topPerProvider =
                    perProvider.entries
                        .sortedByDescending { it.value }
                        .take(MAX_PROVIDERS_PER_DAY)
                        .associate { it.key to it.value }

                val computedTotal = topPerProvider.values.sum()
                val total = dailyPoints[date] ?: computedTotal
                DailySeries(
                    date = date,
                    total = max(total, computedTotal),
                    perProvider = topPerProvider,
                )
            }
        }

        private fun buildHourly(usages: List<TokenUsage>, now: Date): List<HourBucket> {
            val cutoff =
                now.time -
                    MAX_HOURLY_LOOKBACK_DAYS.toLong() *
                    TrendDigestConstants.HOURS_PER_DAY *
                    TrendDigestConstants.SECONDS_PER_MINUTE *
                    TrendDigestConstants.SECONDS_PER_MINUTE *
                    TrendDigestConstants.MILLIS_PER_SECOND
            val cal = Calendar.getInstance()
            val buckets = LongArray(TrendDigestConstants.DEFAULT_SEMANTIC_HASH_LIMIT)
            val costs = DoubleArray(TrendDigestConstants.DEFAULT_SEMANTIC_HASH_LIMIT)
            for (u in usages) {
                if (u.timestamp < cutoff) continue
                cal.time = Date(u.timestamp)
                val h = cal.get(Calendar.HOUR_OF_DAY).coerceIn(0, TrendDigestConstants.MAX_HOUR_OF_DAY)
                buckets[h] = buckets[h] + u.totalTokens
                costs[h] = costs[h] + u.effectiveCost
            }
            return (0..TrendDigestConstants.MAX_HOUR_OF_DAY).map { h ->
                HourBucket(hour = h, costUsd = costs[h], tokens = buckets[h])
            }
        }

        private fun buildSessions(usages: List<TokenUsage>): List<SessionSlice> {
            return usages.take(MAX_SESSIONS).map { u ->
                val duration = ((u.endTime - u.startTime).coerceAtLeast(0) / 1000L).toInt()
                val cacheTotal = u.cacheReadTokens + u.cacheCreationTokens
                val hitRate =
                    if (cacheTotal > 0) {
                        u.cacheReadTokens.toDouble() / cacheTotal.toDouble()
                    } else {
                        0.0
                    }
                val velocity = if (duration > 0) u.outputTokens.toDouble() / duration else 0.0

                SessionSlice(
                    id = u.id.ifBlank { u.sessionId.orEmpty() },
                    startedAtMs = if (u.startTime > 0) u.startTime else u.timestamp,
                    durationSec = duration,
                    provider = AgentProvider.fromKey(u.provider)?.displayName ?: u.provider,
                    providerKey = AgentProvider.fromKey(u.provider)?.key ?: u.provider,
                    model = u.model.orEmpty(),
                    project = u.projectName.orEmpty(),
                    inputTokens = u.inputTokens,
                    outputTokens = u.outputTokens,
                    cacheReadTokens = u.cacheReadTokens,
                    cacheCreationTokens = u.cacheCreationTokens,
                    reasoningTokens = u.reasoningTokens,
                    costUsd = u.effectiveCost,
                    cacheHitRate = hitRate,
                    outputTokensPerSecond = velocity,
                )
            }
        }

        private val ISO_DAY: SimpleDateFormat =
            SimpleDateFormat("yyyy-MM-dd", Locale.US).apply {
                timeZone = TimeZone.getDefault()
            }
    }
}

/** Sentiment tone surfaced by the insight engine. */
enum class TrendInsightTone { POSITIVE, NEUTRAL, WARNING }

/** A single ranked insight rendered by `InsightAutoRotator`. */
data class TrendInsight(
    val id: String,
    val title: String,
    val detail: String,
    val tone: TrendInsightTone,
    /** Material symbol name (Compose `Icons.Filled.X` lookup happens at render time). */
    val symbolName: String,
    val rank: Int,
)

/**
 * Deterministic, stateless rules engine that translates a `TrendDataDigest`
 * into a ranked list of `TrendInsight`s. Ports the iOS
 * `TrendInsightEngine` — same intent, idiomatic Kotlin output.
 */
object TrendInsightEngine {
    fun insights(digest: TrendDataDigest): List<TrendInsight> {
        val today = digest.totals.firstOrNull { it.window == "today" }
        val sevenDay = digest.totals.firstOrNull { it.window == "7d" }
        val thirtyDay = digest.totals.firstOrNull { it.window == "30d" }
        val out = mutableListOf<TrendInsight>()
        trendSpendVelocityInsight(today, sevenDay)?.let { out += it }
        trendProviderConcentrationInsight(digest)?.let { out += it }
        out += trendCacheInsights(digest)
        trendPeakHourInsight(digest)?.let { out += it }
        trendVelocityChampionInsight(digest)?.let { out += it }
        trendLongTailVolumeInsight(thirtyDay)?.let { out += it }
        return out.sortedByDescending { it.rank }
    }
}
